src/Forms/SearchableDropdownTrait.php

```diff
@@ -0,0 +1,594 @@
+     */
+    public function search(HTTPRequest $request): HTTPResponse
+    {
+        $response = HTTPResponse::create();
+        $response->addHeader('Content-Type', 'application/json');
+        if (!SecurityToken::singleton()->checkRequest($request)) {
+            $response->setStatusCode(400);
+            $response->setBody(json_encode(['message' => 'Invalid CSRF token']));
+            return $response;
+        }
+        $term = $request->getVar('term') ?? '';
+        $options = $this->getOptionsForSearchRequest($term);
+        $response->setBody(json_encode($options));
+        return $response;
+    }
+
+    /**
+     * Get whether the currently selected value(s) can be cleared
+     */
+    public function getIsClearable(): bool
+    {
+        return $this->isClearable;
+    }
+
+    /**
+     * Set whether the currently selected value(s) can be cleared
+     */
+    public function setIsClearable(bool $isClearable): static
+    {
+        $this->isClearable = $isClearable;
+        return $this;
+    }
+
+    /**
+     * Get whether values are lazy loading via AJAX
+     */
+    public function getIsLazyLoaded(): bool
+    {
+        return $this->isLazyLoaded;
+    }
+
+    /**
+     * Set whehter values are lazy loaded via AJAX
+     */
+    public function setIsLazyLoaded(bool $isLazyLoaded): static
+    {
+        $this->isLazyLoaded = $isLazyLoaded;
+        if ($isLazyLoaded) {
+            $this->setIsSearchable(true);
+        }
+        return $this;
+    }
+
+    /**
+     * Get the limit of items to lazy load
+     */
+    public function getLazyLoadLimit(): int
+    {
+        return $this->lazyLoadLimit;
+    }
+
+    /**
+     * Set the limit of items to lazy load
+     */
+    public function setLazyLoadLimit(int $lazyLoadLimit): static
+    {
+        $this->lazyLoadLimit = $lazyLoadLimit;
+        return $this;
+    }
+
+    /**
+     * Get the placeholder text
+     */
+    public function getPlaceholder(): string
+    {
+        $placeholder = $this->placeholder;
+        if ($placeholder) {
+            return $placeholder;
+        }
+        // SearchableDropdownField will have the getEmptyString() method from SingleSelectField
+        if (method_exists($this, 'getEmptyString')) {
+            $emptyString = $this->getEmptyString();
+            if ($emptyString) {
+                return $emptyString;
+            }
+        }
+        if ($this->getUseDynamicPlaceholder()) {
+            $parts = [];
+            if (!$this->getIsLazyLoaded()) {
+                $parts[] = _t(__TRAIT__ . '.SELECT', 'select');
+            }
+            if ($this->getIsSearchable()) {
+                $parts[] = _t(__TRAIT__ . '.TYPE_TO_SEARCH', 'type to search');
+            }
+            $conjunctive = _t(__TRAIT__ . '.CONJUNCTIVE', ' or ');
+            $dots = _t(__TRAIT__ . '.DOTS', '...');
+            return ucfirst(implode($conjunctive, $parts)) . $dots;
```

Comment:
Is there a standard translation mechanism for multipart strings with 'or'? Some languages have different ways how to do that than just simply concatenating strings like here.

Reply:
It feels like we're being a little too clever here - and as Michal points out this could cause issues with localisation. I'd say have one or the other (which makes them way easier to localise) - so if it's searchable, just say "Type to search...", and if not, say "Select...".

Though as noted elsewhere, it might not make sense for this to _not_ be searchable, in which case it gets even simpler.

Reply:
Have rewritten to use full sentences rather than joining words together